File: app/src/main/java/com/marked/pixsee/friends/cards/data/MessageDataset.java
package com.marked.pixsee.friends.cards.data;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.marked.pixsee.friends.friends.data.User;
import com.marked.pixsee.data.database.DatabaseContract;
import com.marked.pixsee.data.database.PixyDatabase;
import com.marked.pixsee.data.mapper.MessageToCVMapper;

import org.json.JSONArray;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Created by Tudor Pop on 14-Feb-16.
 */
public class MessageDataset extends ArrayList<Message> {
	private Context mContext;
	private MessageToCVMapper messageToCVMapper = new MessageToCVMapper();

	public MessageDataset(Context mContext) {
		this.mContext = mContext;
	}

	@Override
	public boolean add(Message element) {
		if (element.getMessageType() != MessageConstants.MessageType.TYPING) /* if the message is a 'typing' message, then don't add it to the database */
			PixyDatabase.getInstance(mContext)
					.getWritableDatabase()
					.insertWithOnConflict(DatabaseContract.Message.TABLE_NAME, null, messageToCVMapper.map(element), SQLiteDatabase.CONFLICT_IGNORE);
		return super.add(element);
	}

	@Override
	public boolean addAll(Collection<? extends Message> collection) {
		boolean result = super.addAll(collection);
		{
			SQLiteDatabase database = PixyDatabase.getInstance(mContext).getWritableDatabase();
			database.beginTransaction();
			for (Message message : collection)
				database.insertWithOnConflict(DatabaseContract.Message.TABLE_NAME, null, messageToCVMapper.map(message), SQLiteDatabase.CONFLICT_IGNORE);
			database.endTransaction();
			database.setTransactionSuccessful();
		}
		return result;
	}

	@Override
	public Message set(int index, Message element) {
		PixyDatabase.getInstance(mContext)
				.getWritableDatabase()
				.update(DatabaseContract.Message.TABLE_NAME, messageToCVMapper.map(element), DatabaseContract.Message.COLUMN_ID + " = ?",
						new String[]{element.getId()});
		return super.set(index, element);
	}

	@Override
	public boolean remove(Object element) {
		Message message = (Message) element;
		PixyDatabase.getInstance(mContext).getWritableDatabase().
				                                                        delete(DatabaseContract.Message.TABLE_NAME, DatabaseContract.Message.COLUMN_ID + " = ?s", new String[]{message.getId()});
		return super.remove(element);
	}

	public static MessageDataset messages;

	public static MessageDataset getInstance(Context context) {
		if (messages == null)
			return new MessageDataset(context);
		return messages;
	}

	public JSONArray MessageListToJSONArray(List<Message> list) {
		JSONArray jsonArray = new JSONArray();
		for (Message it : list)
			jsonArray.put(it.toJSON());
		return jsonArray;
	}

	public void loadMore(User friend, int limit) {
		PixyDatabase.getInstance(mContext).getReadableDatabase().beginTransaction();
		Cursor cursor = PixyDatabase.getInstance(mContext).getReadableDatabase().query(
				DatabaseContract.Message.TABLE_NAME,
				new String[]{DatabaseContract.Message.COLUMN_DATA_BODY,
						DatabaseContract.Message.COLUMN_TYPE,
						DatabaseContract.Message.COLUMN_DATE,
						DatabaseContract.Message.COLUMN_TO},
				DatabaseContract.Message.COLUMN_TO + " = " + friend.getUserID(),
				null, null, null, null, String.valueOf(limit)
		);
		cursor.moveToFirst();
		String bodyColName = cursor.getColumnName(cursor.getColumnIndex(DatabaseContract.Message.COLUMN_DATA_BODY));
		String typeColName = cursor.getColumnName(cursor.getColumnIndex(DatabaseContract.Message.COLUMN_TYPE));
		String dateColName = cursor.getColumnName(cursor.getColumnIndex(DatabaseContract.Message.COLUMN_DATE));
		String toColName = cursor.getColumnName(cursor.getColumnIndex(DatabaseContract.Message.COLUMN_TO));
		while (!cursor.isAfterLast()) {
			Message message = new Message.Builder()
					                  .addData(MessageConstants.DATA_BODY, bodyColName)
					                  .messageType(Integer.parseInt(typeColName))
					                  .date(Long.parseLong(dateColName))
					                  .to(toColName).build();
			super.add(message);
		}
	}
}

File: app/src/main/java/com/marked/vifo/model/message/MessageDataset.kt
package com.marked.vifo.model.message

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import com.marked.vifo.extra.MessageConstants
import com.marked.vifo.model.contact.Contact
import com.marked.vifo.model.database.DatabaseContract
import com.marked.vifo.model.database.database
import org.jetbrains.anko.async
import org.jetbrains.anko.db.parseList
import org.jetbrains.anko.db.rowParser
import org.jetbrains.anko.db.select
import org.jetbrains.anko.db.transaction
import org.json.JSONArray
import java.util.*

/**
 * Created by Tudor Pop on 14-Feb-16.
 */
class MessageDataset(var mContext: Context) : ArrayList<Message>() {

	override fun add(element: Message): Boolean {
		if(element.messageType != MessageConstants.MessageType.TYPING) /* if the message is a 'typing' message, then don't add it to the database */
			mContext.database.use {
				insertWithOnConflict(DatabaseContract.Message.TABLE_NAME, null, element.toContentValues(), SQLiteDatabase.CONFLICT_IGNORE)
			}
		return super.add(element)
	}

	override fun addAll(elements: Collection<Message>): Boolean {
		var result = super.addAll(elements)
		async() {
			mContext.database.use {
				transaction {
					elements.forEach {
						insertWithOnConflict(DatabaseContract.Message.TABLE_NAME, null, it.toContentValues(), SQLiteDatabase.CONFLICT_IGNORE)
					}
				}
			}
		}
		return result
	}

	override fun set(index: Int, element: Message): Message {
		mContext.database.use {
			update(DatabaseContract.Message.TABLE_NAME, element.toContentValues(), "${DatabaseContract.Message.COLUMN_ID} = ?s", arrayOf(element.id))
		}
		return super.set(index, element)
	}

	override fun remove(element: Message): Boolean {
		mContext.database.use {
			delete(DatabaseContract.Message.TABLE_NAME, "${DatabaseContract.Message.COLUMN_ID} = ?s", arrayOf(element.id))
		}
		return super.remove(element)
	}

	companion object {
		val messages: MessageDataset? = null
		fun getInstance(context: Context): MessageDataset {
			if (messages == null)
				return MessageDataset(context)
			return messages
		}
	}

	fun MessageListToJSONArray(list: List<Message>): JSONArray {
		val jsonArray = JSONArray()
		for (Message in list)
			jsonArray.put(Message.toJSON())
		return jsonArray
	}

	fun loadMore(contact: Contact, limit: Int = 50) {
		mContext.database.use {
			select(DatabaseContract.Message.TABLE_NAME,
					DatabaseContract.Message.COLUMN_DATA_BODY,
					DatabaseContract.Message.COLUMN_TYPE,
					DatabaseContract.Message.COLUMN_DATE,
					DatabaseContract.Message.COLUMN_TO
			).where("${DatabaseContract.Message.COLUMN_TO}={to}", "to" to contact.id).limit(size, limit)
					.exec {
						parseList(rowParser {
							body: String, type: Int, date: Int, to: String
							->
							val message = Message.Builder().addData(MessageConstants.DATA_BODY, body).messageType(type).date(date.toLong()).to(to).build()
							super.add(message)
						})
						//			}
					}
		}
	}
}